### PR TITLE
Add template for zIPL boot entry option

### PR DESCRIPTION
**Pull request by yuumasato**
#### Description:

- Add `zipl_bls_entries_option` template to check and configure BLS options for zIPL
  - With support for Ansible, Bash and OVAL
- Enables the new template in relevant rules

#### Rationale:
- Templates for the win!

### CODE REVIEW

**Comment by mildas** [comment]
**Changes identified:**
Rule zipl_bls_entries_only:
 OVAL check is newly added.
Rule zipl_bootmap_is_up_to_date:
 OVAL check is newly added.
Profile ospp on rhel8:
 Rule zipl_bootmap_is_up_to_date, zipl_bls_entries_only added to ospp profile.
Profile cui on rhel8:
 CUI profile extends changed OSPP profile.
Profile ospp-mls on rhel8:
 OSPP-MLS profile extends changed OSPP profile.
Profile rhelh-vpp on rhel8:
 RHELH-VPP profile extends changed OSPP profile.
Profile stig on rhel8:
 STIG profile extends changed OSPP profile.
Profile rhelh-stig on rhel8:
 RHELH-STIG profile extends changed STIG profile.
Others:
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 (cd build && cmake ../ && ctest -j4)
 build_product rhel8
 tests/test_suite.py rule --libvirt qemu:///system test-suite-vm --remediate-using bash --datastream build/ssg-rhel8-ds.xml zipl_bls_entries_only
 tests/test_suite.py rule --libvirt qemu:///system test-suite-vm --remediate-using bash --datastream build/ssg-rhel8-ds.xml zipl_bootmap_is_up_to_date
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml cui
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp-mls
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-vpp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml stig

**Comment by mildas** [comment]
**Changes identified:**
Others:
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 (cd build && cmake ../ && ctest -j4)

**Comment by mildas** [comment]
**Changes identified:**
Others:
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 (cd build && cmake ../ && ctest -j4)

**Comment by mildas** [comment]
**Changes identified:**
Profile ospp on rhel8:
 Rule zipl_slub_debug_argument, zipl_page_poison_argument, zipl_vsyscall_argument, zipl_audit_backlog_limit_argument, zipl_pti_argument, zipl_audit_argument added to ospp profile.
Profile cui on rhel8:
 CUI profile extends changed OSPP profile.
Profile ospp-mls on rhel8:
 OSPP-MLS profile extends changed OSPP profile.
Profile rhelh-vpp on rhel8:
 RHELH-VPP profile extends changed OSPP profile.
Profile stig on rhel8:
 STIG profile extends changed OSPP profile.
 Rule !grub2_audit_backlog_limit_argument, !grub2_pti_argument, !grub2_page_poison_argument, !grub2_vsyscall_argument, !grub2_audit_argument, !grub2_slub_debug_argument added to stig profile.
Profile rhelh-stig on rhel8:
 RHELH-STIG profile extends changed STIG profile.
Others:
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 build_product rhel8
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-vpp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml cui
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp-mls
 (cd build && cmake ../ && ctest -j4)

**Comment by mildas** [comment]
**Changes identified:**
Profile ospp on rhel8:
 Rule zipl_vsyscall_argument, zipl_pti_argument, zipl_slub_debug_argument, zipl_page_poison_argument, zipl_audit_argument, zipl_audit_backlog_limit_argument added to ospp profile.
Profile cui on rhel8:
 CUI profile extends changed OSPP profile.
Profile ospp-mls on rhel8:
 OSPP-MLS profile extends changed OSPP profile.
Profile rhelh-vpp on rhel8:
 RHELH-VPP profile extends changed OSPP profile.
Profile stig on rhel8:
 STIG profile extends changed OSPP profile.
 Rule !zipl_audit_backlog_limit_argument, !zipl_vsyscall_argument, !zipl_slub_debug_argument, !zipl_page_poison_argument, !zipl_pti_argument, !zipl_audit_argument added to stig profile.
Profile rhelh-stig on rhel8:
 RHELH-STIG profile extends changed STIG profile.
Others:
 Python abstract syntax tree change found in build-scripts/compile_profiles.py.
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 (cd build && cmake ../ && ctest -j4)
 build_product rhel8
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml cui
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp-mls
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-vpp

**Comment by mildas** [comment]
**Changes identified:**
Profile ospp on rhel8:
 Rule zipl_page_poison_argument, zipl_pti_argument, zipl_audit_argument, zipl_slub_debug_argument, zipl_vsyscall_argument, zipl_audit_backlog_limit_argument added to ospp profile.
Profile cui on rhel8:
 CUI profile extends changed OSPP profile.
Profile ospp-mls on rhel8:
 OSPP-MLS profile extends changed OSPP profile.
Profile rhelh-vpp on rhel8:
 RHELH-VPP profile extends changed OSPP profile.
Profile stig on rhel8:
 Rule !zipl_page_poison_argument, !zipl_vsyscall_argument, !zipl_audit_argument, !zipl_pti_argument, !zipl_audit_backlog_limit_argument, !zipl_slub_debug_argument added to stig profile.
 STIG profile extends changed OSPP profile.
Profile rhelh-stig on rhel8:
 RHELH-STIG profile extends changed STIG profile.
Profile ospp on tests:
 Rule zipl_page_poison_argument, zipl_pti_argument, zipl_audit_argument, zipl_slub_debug_argument, zipl_vsyscall_argument, zipl_audit_backlog_limit_argument added to ospp profile.
Others:
 Python abstract syntax tree change found in build-scripts/compile_profiles.py.
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 build_product tests
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-tests-ds.xml ospp
 (cd build && cmake ../ && ctest -j4)
 build_product rhel8
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-vpp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp-mls
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml cui
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml stig

File: shared/templates/template_OVAL_zipl_bls_entries_option [hunk 1]
```diff
@@ -0,0 +1,47 @@
+<def-group>
+  <definition class="compliance" id="{{{ _RULE_ID }}}" version="1">
+    <metadata>
+      <title>Ensure that BLS-compatible boot loader is configured to run Linux operating system with argument {{{ ARG_NAME_VALUE }}}</title>
+      {{{- oval_affected(products) }}}
+      <description>Ensure {{{ ARG_NAME_VALUE }}} option is configured in the 'options' line in /boot/loader/entries/*.conf.</description>
+    </metadata>
+    <criteria operator="AND">
+      <criterion test_ref="test_bls_{{{ SANITIZED_ARG_NAME }}}_options"
+      comment="Check if {{{ ARG_NAME_VALUE }}} is present in the 'options' line in /boot/loader/entries/*" />
```
Comment (vojtapolasek) [inline shared/templates/template_OVAL_zipl_bls_entries_option:10]:
```suggestion
      comment="Check if {{{ ARG_NAME_VALUE }}} is present in the 'options' line in /boot/loader/entries/*.conf" />
```

**Comment by mildas** [comment]
**Changes identified:**
Profile ospp on rhel8:
 Rule zipl_vsyscall_argument, zipl_audit_argument, zipl_pti_argument, zipl_slub_debug_argument, zipl_page_poison_argument, zipl_audit_backlog_limit_argument added to ospp profile.
Profile cui on rhel8:
 CUI profile extends changed OSPP profile.
Profile stig on rhel8:
 STIG profile extends changed OSPP profile.
 Rule !zipl_vsyscall_argument, !zipl_audit_backlog_limit_argument, !zipl_pti_argument, !zipl_page_poison_argument, !zipl_slub_debug_argument, !zipl_audit_argument added to stig profile.
Profile rhelh-stig on rhel8:
 RHELH-STIG profile extends changed STIG profile.
Profile rhelh-vpp on rhel8:
 RHELH-VPP profile extends changed OSPP profile.
Profile ospp-mls on rhel8:
 OSPP-MLS profile extends changed OSPP profile.
Profile ospp on tests:
 Rule zipl_vsyscall_argument, zipl_audit_argument, zipl_pti_argument, zipl_slub_debug_argument, zipl_page_poison_argument, zipl_audit_backlog_limit_argument added to ospp profile.
Others:
 Python abstract syntax tree change found in build-scripts/compile_profiles.py.
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 (cd build && cmake ../ && ctest -j4)
 build_product tests
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-tests-ds.xml ospp
 build_product rhel8
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml cui
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-vpp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp-mls

File: shared/templates/template_BASH_zipl_bls_entries_option [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+# Correct BLS option using grubby, which is a thin wrapper around BLS operations
+grubby --update-kernel=ALL --args="{{{ ARG_NAME_VALUE }}}"
+
+# Ensure new kernels and boot entries retain the boot option
+if [ ! -f /etc/kernel/cmdline ]; then
+    echo "{{{ ARG_NAME_VALUE }}}" >> /etc/kernel/cmdline
+elif ! grep -q '^(.*\s)?{{{ ARG_NAME_VALUE }}}(\s.*)?$' /etc/kernel/cmdline; then
+    echo " audit=1" >> /etc/kernel/cmdline
+    sed -Ei 's/^(.*)$/\1 audit=1/' /etc/kernel/cmdline
```
Comment (matejak) [inline shared/templates/template_BASH_zipl_bls_entries_option:11]: What is this `audit=1` thing?

[the remark this answers is not in the document]
Comment (yuumasato) [reply shared/templates/template_BASH_zipl_bls_entries_option:11]: Achjo, this is becoming a classic mistake in templates, :D

File: linux_os/guide/system/bootloader-zipl/zipl_audit_argument/tests/correct_option.pass.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+# platform = Red Hat Enterprise Linux 8
```
Comment (matejak) [inline linux_os/guide/system/bootloader-zipl/zipl_audit_argument/tests/correct_option.pass.sh:2]: I wouldn't be so strict with the platform - it is a textfilecontent test, so actually any platform will do. The remediation requires `grubby`, but that can be installed into Fedora as well.

**Comment by vojtapolasek** [comment]
Thank you for tests. However, please count on the case that /etc/kernel/cmdline does not exist before test is run.

File: linux_os/guide/system/bootloader-zipl/zipl_audit_argument/tests/correct_option.pass.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+# remediation = none
+
+# Make sure boot loader entries contain audit=1
+for file in /boot/loader/entries/*.conf
+do
+    if ! grep -q '^options.*audit=1.*$' "$file" ; then
+        sed -i '/^options / s/$/audit=1/' "$file"
```
Comment (vojtapolasek) [inline linux_os/guide/system/bootloader-zipl/zipl_audit_argument/tests/correct_option.pass.sh:9]:
```suggestion
        sed -i '/^options / s/$/ audit=1/' "$file"
```

File: shared/templates/template_BASH_zipl_bls_entries_option (hunk 2 above)
Comment (yuumasato) [reply shared/templates/template_BASH_zipl_bls_entries_option:11]: Mistake when converting a specific remediation to a template, :laughing:

**Comment by mildas** [comment]
**Changes identified:**
Profile ospp on rhel8:
 Rule zipl_audit_backlog_limit_argument, zipl_vsyscall_argument, zipl_audit_argument, zipl_pti_argument, zipl_slub_debug_argument, zipl_page_poison_argument added to ospp profile.
Profile cui on rhel8:
 CUI profile extends changed OSPP profile.
Profile stig on rhel8:
 Rule !zipl_pti_argument, !zipl_audit_argument, !zipl_slub_debug_argument, !zipl_audit_backlog_limit_argument, !zipl_page_poison_argument, !zipl_vsyscall_argument added to stig profile.
 STIG profile extends changed OSPP profile.
Profile rhelh-stig on rhel8:
 RHELH-STIG profile extends changed STIG profile.
Profile rhelh-vpp on rhel8:
 RHELH-VPP profile extends changed OSPP profile.
Profile ospp-mls on rhel8:
 OSPP-MLS profile extends changed OSPP profile.
Profile ospp on tests:
 Rule zipl_audit_backlog_limit_argument, zipl_vsyscall_argument, zipl_audit_argument, zipl_pti_argument, zipl_slub_debug_argument, zipl_page_poison_argument added to ospp profile.
Others:
 Python abstract syntax tree change found in build-scripts/compile_profiles.py.
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 build_product rhel8
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml cui
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-vpp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp-mls
 (cd build && cmake ../ && ctest -j4)
 build_product tests
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-tests-ds.xml ospp

File: shared/templates/template_ANSIBLE_zipl_bls_entries_option [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      find:
+        paths: "/boot/loader/entries/"
+        contains: "^options .*{{{ ARG_NAME_VALUE }}}.*$"
+        patterns: "*.conf"
+      register: entries_options
+
+    - name: "Update boot entries options"
+      command: grubby --update-kernel=ALL --args="{{{ ARG_NAME_VALUE }}}"
+      when: entries_options is defined and entries_options.examined != entries_options.matched
+      # The conditional above assumes that only *.conf files are present in /boot/loader/entries
```
[the remark this answers is not in the document]
Comment (yuumasato) [reply shared/templates/template_ANSIBLE_zipl_bls_entries_option]: I updated the task to count number of`conf` files.

**Comment by mildas** [comment]
**Changes identified:**
Profile ospp on rhel8:
 Rule zipl_audit_argument, zipl_slub_debug_argument, zipl_pti_argument, zipl_page_poison_argument, zipl_audit_backlog_limit_argument, zipl_vsyscall_argument added to ospp profile.
Profile cui on rhel8:
 CUI profile extends changed OSPP profile.
Profile stig on rhel8:
 STIG profile extends changed OSPP profile.
 Rule !zipl_audit_argument, !zipl_slub_debug_argument, !zipl_audit_backlog_limit_argument, !zipl_pti_argument, !zipl_page_poison_argument, !zipl_vsyscall_argument added to stig profile.
Profile rhelh-stig on rhel8:
 RHELH-STIG profile extends changed STIG profile.
Profile rhelh-vpp on rhel8:
 RHELH-VPP profile extends changed OSPP profile.
Profile ospp-mls on rhel8:
 OSPP-MLS profile extends changed OSPP profile.
Profile ospp on tests:
 Rule zipl_audit_argument, zipl_slub_debug_argument, zipl_pti_argument, zipl_page_poison_argument, zipl_audit_backlog_limit_argument, zipl_vsyscall_argument added to ospp profile.
Others:
 Python abstract syntax tree change found in build-scripts/compile_profiles.py.
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 (cd build && cmake ../ && ctest -j4)
 build_product tests
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-tests-ds.xml ospp
 build_product rhel8
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml cui
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp-mls
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-vpp

File: linux_os/guide/system/bootloader-zipl/zipl_audit_argument/tests/correct_option.pass.sh (hunk 3 above)
Comment (yuumasato) [reply linux_os/guide/system/bootloader-zipl/zipl_audit_argument/tests/correct_option.pass.sh]: I enabled it for Fedora as well, and removed `# remediation = none`.

**Comment by yuumasato** [comment]
@vojtapolasek @matejak Thank you for the review, gentleman. Hopefully all the issues have been addressed.

**Comment by vojtapolasek** [comment]
One last thing. Currently, when running tests, if /etc/kernel/cmdline does not exist, the grep gives an error because it can't grep... I don't know if it needs some fixing. Maybe I discovered it cause I was running tests manually.

**Comment by yuumasato** [comment]
> One last thing. Currently, when running tests, if /etc/kernel/cmdline does not exist, the grep gives an error because it can't grep... I don't know if it needs some fixing. Maybe I discovered it cause I was running tests manually.

`grep` can suppress error messages with `-s`.. I could add that.

**Comment by mildas** [comment]
**Changes identified:**
Profile ospp on rhel8:
 Rule zipl_audit_argument, zipl_audit_backlog_limit_argument, zipl_page_poison_argument, zipl_vsyscall_argument, zipl_pti_argument, zipl_slub_debug_argument added to ospp profile.
Profile cui on rhel8:
 CUI profile extends changed OSPP profile.
Profile stig on rhel8:
 STIG profile extends changed OSPP profile.
 Rule !zipl_vsyscall_argument, !zipl_slub_debug_argument, !zipl_page_poison_argument, !zipl_audit_backlog_limit_argument, !zipl_audit_argument, !zipl_pti_argument added to stig profile.
Profile rhelh-stig on rhel8:
 RHELH-STIG profile extends changed STIG profile.
Profile rhelh-vpp on rhel8:
 RHELH-VPP profile extends changed OSPP profile.
Profile ospp-mls on rhel8:
 OSPP-MLS profile extends changed OSPP profile.
Profile ospp on tests:
 Rule zipl_audit_argument, zipl_audit_backlog_limit_argument, zipl_page_poison_argument, zipl_vsyscall_argument, zipl_pti_argument, zipl_slub_debug_argument added to ospp profile.
Others:
 Python abstract syntax tree change found in build-scripts/compile_profiles.py.
 Python abstract syntax tree change found in ssg/templates.py.

**Recommended tests to execute:**
 build_product tests
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-tests-ds.xml ospp
 build_product rhel8
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml cui
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp-mls
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-stig
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml rhelh-vpp
 tests/test_suite.py profile --libvirt qemu:///system test-suite-vm --datastream build/ssg-rhel8-ds.xml ospp
 (cd build && cmake ../ && ctest -j4)

**Comment by vojtapolasek** [comment]
Thank you very much for this template. Merging.

**Pull request merged**